test/hotspot/jtreg/serviceability/jvmti/RedefineClasses/RedefineSharedClassJFR.java

```diff
@@ -0,0 +1,102 @@
+        // on if the test is run with or without CDS the output will be different,
+        // due to the fact that shared classes can never be cleaned out after retranform.
+        if (args.length > 0) {
+            // When run with an argument the class is used as driver and should parse
+            // the output.
+            if (args[0].equals("xshare-off")) {
+                // First case is with -Xshare:off. In this case no classes are shared
+                // and we should be able to clean out the retransformed classes. Verify
+                // that the cleaning is done when the GC is triggered.
+                ProcessBuilder pb = ProcessTools.createJavaProcessBuilder(
+                    "-XX:StartFlightRecording",
+                    "-Xshare:off",
+                    "-Xlog:redefine+class+iklass+add=trace,redefine+class+iklass+purge=trace",
+                    "RedefineSharedClassJFR");
+                new OutputAnalyzer(pb.start())
+                    .shouldContain("Class unloading: clean_previous_versions = true")
+                    .shouldNotContain("Class unloading: clean_previous_versions = false")
+                    // We expect at least one of the transformed classes to be in use, if
+                    // not the above check that clean_previous should be true will also
+                    // fail. This check is to show what is expected.
+                    .shouldContain("scratch class added; one of its methods is on_stack.")
+                    .shouldNotContain("scratch class added; class is shared")
+                    .shouldHaveExitValue(0);
+                return;
+            } else if (args[0].equals("xshare-on")) {
+                // With -Xshare:on, the shared classes can never be cleaned out. Check the
+                // logs to verify we don't try to clean when we know it is not needed.
+                ProcessBuilder pb = ProcessTools.createJavaProcessBuilder(
+                    "-XX:StartFlightRecording",
+                    "-Xshare:on",
+                    "-Xlog:redefine+class+iklass+add=trace,redefine+class+iklass+purge=trace",
+                    "RedefineSharedClassJFR");
+                new OutputAnalyzer(pb.start())
+                    .shouldContain("Class unloading: clean_previous_versions = false")
+                    .shouldNotContain("Class unloading: clean_previous_versions = true")
+                    .shouldContain("scratch class added; class is shared")
+                    // If the below line occurs, then clean_previous_versions will be
+                    // true and the above shouldContain will trigger. This check is to
+                    // show the intention that we don't expect any non-shared transformed
+                    // classes to be in use.
+                    .shouldNotContain("scratch class added; one of its methods is on_stack.")
+                    .shouldHaveExitValue(0);
+                return;
```

Comment:
The fragments 61-74 and 79-93 have a big common part which can be a good base for a refactoring.
But it can be not worth it. So, I leave it up to you.